File: Models/BiomaSurfacePartonSoilSWATHourlyPartonC/src/simplace/BiomaSurfacePartonSoilSWATHourlyPartonC/RangeOfSoilTemperaturesDAYCENT.java
package net.simplace.sim.components.BiomaSurfacePartonSoilSWATHourlyPartonC;
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom2.Element;


public class RangeOfSoilTemperaturesDAYCENT extends FWSimComponent
{
    private FWSimVariable<Double[]> LayerThickness;
    private FWSimVariable<Double> SurfaceTemperatureMinimum;
    private FWSimVariable<Double[]> ThermalDiffusivity;
    private FWSimVariable<Double[]> SoilTemperatureByLayers;
    private FWSimVariable<Double> SurfaceTemperatureMaximum;
    private FWSimVariable<Double[]> SoilTemperatureRangeByLayers;
    private FWSimVariable<Double[]> SoilTemperatureMinimum;
    private FWSimVariable<Double[]> SoilTemperatureMaximum;

    public RangeOfSoilTemperaturesDAYCENT(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public RangeOfSoilTemperaturesDAYCENT(){
        super();
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("LayerThickness", "Soil layer thickness", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.constant,"m", 0.005, 3, 0.05, this));
        addVariable(FWSimVariable.createSimVariable("SurfaceTemperatureMinimum", "Minimum surface soil temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"degC", -60, 60, 10, this));
        addVariable(FWSimVariable.createSimVariable("ThermalDiffusivity", "Thermal diffusivity of soil layer", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.input,"mm s-1", 0, 1, 0.0025, this));
        addVariable(FWSimVariable.createSimVariable("SoilTemperatureByLayers", "Soil temperature of each layer", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.input,"degC", -60, 60, 15, this));
        addVariable(FWSimVariable.createSimVariable("SurfaceTemperatureMaximum", "Maximum surface soil temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"degC", -60, 60, 25, this));
        addVariable(FWSimVariable.createSimVariable("SoilTemperatureRangeByLayers", "Soil temperature range by layers", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.out,"degC", 0, 50, null, this));
        addVariable(FWSimVariable.createSimVariable("SoilTemperatureMinimum", "Minimum soil temperature by layers", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.out,"", -60, 60, null, this));
        addVariable(FWSimVariable.createSimVariable("SoilTemperatureMaximum", "Maximum soil temperature by layers", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.out,"degC", -60, 60, null, this));

        return iFieldMap;
    }
    @Override
    protected void process()
    {
        Double [] t_LayerThickness = LayerThickness.getValue();
        Double t_SurfaceTemperatureMinimum = SurfaceTemperatureMinimum.getValue();
        Double [] t_ThermalDiffusivity = ThermalDiffusivity.getValue();
        Double [] t_SoilTemperatureByLayers = SoilTemperatureByLayers.getValue();
        Double t_SurfaceTemperatureMaximum = SurfaceTemperatureMaximum.getValue();
        Double [] t_SoilTemperatureRangeByLayers;
        Double [] t_SoilTemperatureMinimum;
        Double [] t_SoilTemperatureMaximum;
        Integer i;
        Double _DepthBottom;
        Double _DepthCenterLayer;
        Double SurfaceDiurnalRange;
        _DepthBottom = (double)(0);
        _DepthCenterLayer = (double)(0);
        SurfaceDiurnalRange = t_SurfaceTemperatureMaximum - t_SurfaceTemperatureMinimum;
        for (i=0 ; i!=t_SoilTemperatureByLayers.length ; i+=1)
        {
            if (i == 0)
            {
                _DepthCenterLayer = t_LayerThickness[0] * 100 / 2;
                t_SoilTemperatureRangeByLayers[0] = SurfaceDiurnalRange * Math.exp(-_DepthCenterLayer * Math.pow(0.00005d / t_ThermalDiffusivity[0], 0.5d));
                t_SoilTemperatureMaximum[0] = t_SoilTemperatureByLayers[0] + (t_SoilTemperatureRangeByLayers[0] / 2);
                t_SoilTemperatureMinimum[0] = t_SoilTemperatureByLayers[0] - (t_SoilTemperatureRangeByLayers[0] / 2);
            }
            else
            {
                _DepthBottom = _DepthBottom + (t_LayerThickness[(i - 1)] * 100);
                _DepthCenterLayer = _DepthBottom + (t_LayerThickness[i] * 100 / 2);
                t_SoilTemperatureRangeByLayers[i] = SurfaceDiurnalRange * Math.exp(-_DepthCenterLayer * Math.pow(0.00005d / t_ThermalDiffusivity[i], 0.5d));
                t_SoilTemperatureMaximum[i] = t_SoilTemperatureByLayers[i] + (t_SoilTemperatureRangeByLayers[i] / 2);
                t_SoilTemperatureMinimum[i] = t_SoilTemperatureByLayers[i] - (t_SoilTemperatureRangeByLayers[i] / 2);
            }
        }
        SoilTemperatureRangeByLayers.setValue(t_SoilTemperatureRangeByLayers, this);
        SoilTemperatureMinimum.setValue(t_SoilTemperatureMinimum, this);
        SoilTemperatureMaximum.setValue(t_SoilTemperatureMaximum, this);
    }

    @Override
    protected void init()
    {
    }
    public HashMap<String, FWSimVariable<?>> fillTestVariables(int aParamIndex, TEST_STATE aDefineOrCheck)
    {
        return iFieldMap;
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new RangeOfSoilTemperaturesDAYCENT(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }
}

File: Models/BiomaSurfacePartonSoilSWATHourlyPartonC/src/apsim/BiomaSurfacePartonSoilSWATHourlyPartonC/RangeOfSoilTemperaturesDAYCENT.cs
using System;
using System.Collections.Generic;
using System.Linq;    
using Models.Core;   
namespace Models.Crop2ML;

/// <summary>
///- Name: RangeOfSoilTemperaturesDAYCENT -Version: 001, -Time step: 1
///- Description:
///            * Title: RangeOfSoilTemperaturesDAYCENT model
///            * Authors: simone.bregaglio
///            * Reference: http://bioma.jrc.ec.europa.eu/ontology/JRC_MARS_biophysical_domain.owl
///            * Institution: University Of Milan
///            * ExtendedDescription: Strategy for the calculation of soil thermal conductivity.Reference: DAYCENT model written in C code
///            * ShortDescription: Strategy for the calculation of soil thermal conductivity
///- inputs:
///            * name: LayerThickness
///                          ** description : Soil layer thickness
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLEARRAY
///                          ** len : 
///                          ** max : 3
///                          ** min : 0.005
///                          ** default : 0.05
///                          ** unit : m
///            * name: SurfaceTemperatureMinimum
///                          ** description : Minimum surface soil temperature
///                          ** inputtype : variable
///                          ** variablecategory : auxiliary
///                          ** datatype : DOUBLE
///                          ** max : 60
///                          ** min : -60
///                          ** default : 10
///                          ** unit : degC
///            * name: ThermalDiffusivity
///                          ** description : Thermal diffusivity of soil layer
///                          ** inputtype : variable
///                          ** variablecategory : auxiliary
///                          ** datatype : DOUBLEARRAY
///                          ** len : 
///                          ** max : 1
///                          ** min : 0
///                          ** default : 0.0025
///                          ** unit : mm s-1
///            * name: SoilTemperatureByLayers
///                          ** description : Soil temperature of each layer
///                          ** inputtype : variable
///                          ** variablecategory : auxiliary
///                          ** datatype : DOUBLEARRAY
///                          ** len : 
///                          ** max : 60
///                          ** min : -60
///                          ** default : 15
///                          ** unit : degC
///            * name: SurfaceTemperatureMaximum
///                          ** description : Maximum surface soil temperature
///                          ** inputtype : variable
///                          ** variablecategory : auxiliary
///                          ** datatype : DOUBLE
///                          ** max : 60
///                          ** min : -60
///                          ** default : 25
///                          ** unit : degC
///            * name: SoilTemperatureRangeByLayers
///                          ** description : Soil temperature range by layers
///                          ** inputtype : variable
///                          ** variablecategory : state
///                          ** datatype : DOUBLEARRAY
///                          ** len : 
///                          ** max : 50
///                          ** min : 0
///                          ** default : 
///                          ** unit : degC
///            * name: SoilTemperatureMinimum
///                          ** description : Minimum soil temperature by layers
///                          ** inputtype : variable
///                          ** variablecategory : state
///                          ** datatype : DOUBLEARRAY
///                          ** len : 
///                          ** max : 60
///                          ** min : -60
///                          ** default : 
///                          ** unit : degC
///            * name: SoilTemperatureMaximum
///                          ** description : Maximum soil temperature by layers
///                          ** inputtype : variable
///                          ** variablecategory : state
///                          ** datatype : DOUBLEARRAY
///                          ** len : 
///                          ** max : 60
///                          ** min : -60
///                          ** default : 
///                          ** unit : degC
///- outputs:
///            * name: SoilTemperatureRangeByLayers
///                          ** description : Soil temperature range by layers
///                          ** datatype : DOUBLEARRAY
///                          ** variablecategory : state
///                          ** len : 
///                          ** max : 50
///                          ** min : 0
///                          ** unit : degC
///            * name: SoilTemperatureMinimum
///                          ** description : Minimum soil temperature by layers
///                          ** datatype : DOUBLEARRAY
///                          ** variablecategory : state
///                          ** len : 
///                          ** max : 60
///                          ** min : -60
///                          ** unit : degC
///            * name: SoilTemperatureMaximum
///                          ** description : Maximum soil temperature by layers
///                          ** datatype : DOUBLEARRAY
///                          ** variablecategory : state
///                          ** len : 
///                          ** max : 60
///                          ** min : -60
///                          ** unit : degC
/// </summary>
public class RangeOfSoilTemperaturesDAYCENT
{

    /// <summary>
    /// initialization of the RangeOfSoilTemperaturesDAYCENT component
    /// </summary>
    public void Init(SurfacePartonSoilSWATHourlyPartonCState s, SurfacePartonSoilSWATHourlyPartonCState s1, SurfacePartonSoilSWATHourlyPartonCRate r, SurfacePartonSoilSWATHourlyPartonCAuxiliary a, SurfacePartonSoilSWATHourlyPartonCExogenous ex)
    {
        double[] SoilTemperatureRangeByLayers ;
        double[] SoilTemperatureMinimum ;
        double[] SoilTemperatureMaximum ;
        SoilTemperatureRangeByLayers = new double[LayerThickness.Length];
        SoilTemperatureMaximum = new double[LayerThickness.Length];
        SoilTemperatureMinimum = new double[LayerThickness.Length];
        s.SoilTemperatureRangeByLayers= SoilTemperatureRangeByLayers;
        s.SoilTemperatureMinimum= SoilTemperatureMinimum;
        s.SoilTemperatureMaximum= SoilTemperatureMaximum;
    }

    private double[] _LayerThickness;
    /// <summary>
    /// Gets and sets the Soil layer thickness
    /// </summary>
    [Description("Soil layer thickness")] 
    [Units("m")] 
    //[Crop2ML(datatype="DOUBLEARRAY", min=0.005, max=3, default=0.05, parametercategory=constant, inputtype="parameter")] 
    public double[] LayerThickness
    {
        get { return this._LayerThickness; }
        set { this._LayerThickness= value; } 
    }

    
    /// <summary>
    /// Constructor of the RangeOfSoilTemperaturesDAYCENT component")
    /// </summary>  
    public RangeOfSoilTemperaturesDAYCENT() { }
    
    /// <summary>
    /// Algorithm of the RangeOfSoilTemperaturesDAYCENT component
    /// </summary>
    public void  CalculateModel(SurfacePartonSoilSWATHourlyPartonCState s, SurfacePartonSoilSWATHourlyPartonCState s1, SurfacePartonSoilSWATHourlyPartonCRate r, SurfacePartonSoilSWATHourlyPartonCAuxiliary a, SurfacePartonSoilSWATHourlyPartonCExogenous ex)
    {
        double SurfaceTemperatureMinimum = a.SurfaceTemperatureMinimum;
        double[] ThermalDiffusivity = a.ThermalDiffusivity;
        double[] SoilTemperatureByLayers = a.SoilTemperatureByLayers;
        double SurfaceTemperatureMaximum = a.SurfaceTemperatureMaximum;
        double[] SoilTemperatureRangeByLayers = s.SoilTemperatureRangeByLayers;
        double[] SoilTemperatureMinimum = s.SoilTemperatureMinimum;
        double[] SoilTemperatureMaximum = s.SoilTemperatureMaximum;
        int i;
        double _DepthBottom;
        double _DepthCenterLayer;
        double SurfaceDiurnalRange;
        _DepthBottom = (double)(0);
        _DepthCenterLayer = (double)(0);
        SurfaceDiurnalRange = SurfaceTemperatureMaximum - SurfaceTemperatureMinimum;
        for (i=0 ; i!=SoilTemperatureByLayers.Length ; i+=1)
        {
            if (i == 0)
            {
                _DepthCenterLayer = LayerThickness[0] * 100 / 2;
                SoilTemperatureRangeByLayers[0] = SurfaceDiurnalRange * Math.Exp(-_DepthCenterLayer * Math.Pow(0.00005 / ThermalDiffusivity[0], 0.5));
                SoilTemperatureMaximum[0] = SoilTemperatureByLayers[0] + (SoilTemperatureRangeByLayers[0] / 2);
                SoilTemperatureMinimum[0] = SoilTemperatureByLayers[0] - (SoilTemperatureRangeByLayers[0] / 2);
            }
            else
            {
                _DepthBottom = _DepthBottom + (LayerThickness[(i - 1)] * 100);
                _DepthCenterLayer = _DepthBottom + (LayerThickness[i] * 100 / 2);
                SoilTemperatureRangeByLayers[i] = SurfaceDiurnalRange * Math.Exp(-_DepthCenterLayer * Math.Pow(0.00005 / ThermalDiffusivity[i], 0.5));
                SoilTemperatureMaximum[i] = SoilTemperatureByLayers[i] + (SoilTemperatureRangeByLayers[i] / 2);
                SoilTemperatureMinimum[i] = SoilTemperatureByLayers[i] - (SoilTemperatureRangeByLayers[i] / 2);
            }
        }
        s.SoilTemperatureRangeByLayers= SoilTemperatureRangeByLayers;
        s.SoilTemperatureMinimum= SoilTemperatureMinimum;
        s.SoilTemperatureMaximum= SoilTemperatureMaximum;
    }
}